FIX the SELECT examine more than MAX_JOIN_SIZE rows #12305

olibre opened the pull request:
# Fix #12305 the SELECT examine more than MAX_JOIN_SIZE rows

Mon [hébergement mutualisé](https://www.online.net/fr/hosting/online-perso) limite `MAX_JOIN_SIZE` :

```sql
SELECT @@MAX_JOIN_SIZE
300000
```

Sur mon ordinateur personnel, j'ai une valeur immense :

```sql
SELECT @@MAX_JOIN_SIZE
18446744073709551615
```

La correction consiste à prévenir MySQL que c'est un gros `SELECT` avec les commandes suivantes :

```php
$db->query("SET SQL_BIG_SELECTS=1");

[... le code du gros SELECT ...]

$db->query("SET SQL_BIG_SELECTS=0");
```

Je n'ai pas testé avec PostgreSQL.

olibre committed:
FIX the SELECT examine more than MAX_JOIN_SIZE rows #12305

eldy commented on htdocs/accountancy/customer/index.php:
```diff
@@ -140,6 +140,9 @@
 	dol_syslog('htdocs/accountancy/customer/index.php');
+	// MAX_JOIN_SIZE can be very low (ex: 300000) on some limited configurations (ex: https://www.online.net/fr/hosting/online-perso)
+	// This big SELECT command may exceed the MAX_JOIN_SIZE limit => Therefore we use SQL_BIG_SELECTS=1 to disable the MAX_JOIN_SIZE security
+	$db->query("SET SQL_BIG_SELECTS=1");
```
You should add here a test 
if ($db->type == 'mysqli') so the code will be done only for mysql

olibre committed:
Check if mysqli to avoid perturbing PostgreSQL with "SET SQL_BIG_SELECTS=1"

olibre commented:
Thanks for your advice, I have just added the check `if ($db->type == 'mysqli')` :smile: 

I have also removed the unnecessary `SET SQL_BIG_SELECTS=0` in `htdocs/accountancy/customer/index.php`. It still works on my limited shared hosting. I did a `git push --force` to avoid polluting Dolibarr Git history.

I am not expert about SQL, but according my understanding, the statement `SET SQL_BIG_SELECTS=0` after the `SELECT` is not required because `SQL_BIG_SELECTS=1` seems to apply to only one `SELECT`. However, in case of misunderstanding, I prefer to keep it. What do you think?
